### ISSUE TITLE
Interchangeability of AtomGroup and selection strings

### ISSUE THREAD

**Issue by GoogleCodeExporter** [MDAnalysis/mdanalysis#175]
Following the discussion on Issue #171 it would be interesting to adapt any functions that take AtomGroups to also be able to take selection strings (and vice versa).

I brought it up because an analysis class I use often, `LeafletFinder`, accepts exclusively selection strings. I have a fix for this particular case, and since I'm working on Issue #174 I'll look into allowing `AtomGroups` to be passed to selection functions.

Original issue reported on code.google.com by `manuel.n...@gmail.com` on 28 Mar 2014 at 1:16

**Comment by mnmelo** [comment]
Ok, with #719 being raised, I decided to tackle this once and for all.
I need feedback on the API implementation, and how it should blend with the analysis API decisions in #719.

tl;dr: We extend `AtomGroup` with an `updating` attribute. Users set it either via `select_atoms` or after an `AtomGroup` is created. `updating` can be a selection string or a callback.

The problem
----------------
Some analyses take universes and selection strings as arguments, others AtomGroups, and yet others just coordinate arrays (although we might admit this if they're low-level routines, like `distance_array`). This is inconsistent.

I can't come up with a clean duck-typing approach where we'd try to use an argument as an `AtomGroup` and fallback to treating it as a selection string on failure. It'd be very messy to have `**kwargs` overloading on every analysis to deal with both cases.

Enforcing that analyses only accept `AtomGroups` is inflexible. In particular, as @richardjgowers pointed out, an analysis might expect a selection string so that the selected group can be updated for every analyzed frame.

What I propose
--------------------
We extend the `AtomGroup` class to implement a caching mechanism so that a selection is reevaluated if the frame has changed. This mechanism is off by default, and only kicks in if `AtomGroup.updating` is set. This allows for an `AtomGroup` to be used in a context that requires selection updates.

For greater flexibility and complex selections `updating` can also accept a callback instead of a selection string. The callback must return an `AtomGroup`. (For consistency with `select_atoms` updating can also be a tuple of selection strings/callbacks, and the result is the combined `AtomGroup`.)

Examples
-------------
```python
ag = universe.select_atoms("resname SOL and around 10 protein", updating=True)

# or post-facto, if updating was left at False (its default):
ag.updating = "resname SOL and around 10 protein"

# also with a callback:
def get_top_leaflet():
    # an example of a complex selection based on leaflet assignment
    top, bottom = MDAnalysis.analysis.leaflet.LeafletFinder...
    return top
ag = universe.select_atoms(get_top_leaflet, updating=True)

# can also be switched off    
ag = universe.updating = None
```

Problems
------------
- kwargs in `select_atoms` are already used by named groups in selections (#174). To implement my suggestion we must create a list of reserved words that can't be used for named groups.
- Passing selections with named groups will not reflect updates to those groups, unless they are themselves also set to update:
```python
top_atoms = universe.select_atoms("prop z > 40")
top_solvating = universe.select_atoms("name SOL and around 10 group top",
        top=top_atoms, updating=True)

# top_solvating keeps the reference to the initial top_atoms group,
#  even when some atoms move below z = 40.
# Reevaluating top_atoms won't help:
top_atoms = universe.select_atoms("prop z > 40")
# top_solvating still holds a reference to the old top_atoms group
#  there's no way it can know the top_atoms name was reassigned.

# The solution is to also set top_atoms to auto update:
top_atoms = universe.select_atoms("prop z > 40", updating=True)
top_solvating = universe.select_atoms("name SOL and around 10 group top",
        top=top_atoms, updating=True)
```

**Comment by jbarnoud** [comment]
I like this proposal very much; I was thinking about a more convoluted solution, this one is much better.

> kwargs in `select_atoms` are already used by named groups in selections (#174). To implement my suggestion we must create a list of reserved words that can't be used for named groups.

I do not think it is a big deal to have *one* reserved keyword. The odd of an `AtomGroup` being called `updating` are rather low any way.

> Passing selections with named groups will not reflect updates to those groups, unless they are themselves also set to update:

It is not a bug, it's a feature! I do not want a selection to be updated without my knowledge.

**Comment by richardjgowers** [comment]
Hmm, I'm not sure I like making AtomGroup more complicated.  How about a new class, like AtomSelection, so

```python
atomsel = mda.AtomSelection(u, 'prop z > 40')

# Automatically update?
for ts in u.trajectory:
    # Would be different each time
    print len(atomsel)

# Or make it manual?
for ts in u.trajectory:
    atomsel.refresh()
    print len(atomsel)
```

I'm not sure how to make the object automatically update, but maybe just hacking into `getattr` will be enough

```python
class AtomSelection(AtomGroup):
    def __getattr__(self, attr):
        if not self.universe.ts.frame = self._my_frame:
            self.refresh()
        return super().__getattr__(attr)
```

Finally, with the way selections work, it should be possible to do this using the `Selection` objects somehow, to avoid having to go through the Parser twice.

**Comment by dotsdl** [comment]
I agree with @richardjgowers suggestion. In the topology overhaul for #363 `AtomGroup` is immutable and light, with the only state it carries being its array of atom indices. This proposal would make this rather complicated on first impression.

**Comment by mnmelo** [comment]
Cool idea about saving selection objects, @richardjgowers . That should speed up things.

I feel that incorporating this functionality in `AtomGroup` will be more user-friendly.
But I agree it will add some weight to `AtomGroup`. At first I also thought about a different class, which would be essentially a proxy for an `AtomGroup`, with a thin caching and selection layer. The downside is that this would force users to be aware of the difference between the two classes, whereas overloading `AtomGroup` would be transparent.

Here's what I foresee would be affected in `AtomGroup`. If you feel it's too much I'll focus on the separate-class implementation:
- Upon creation the object would save the selection objects/callbacks, and possibly a reference to the creating `AtomGroup` (for the cases where `select_atoms` is invoked from an `AtomGroup` instead of from a `Universe`);
- The object would keep a `weakref` to the `Timestep` of last refresh;
- Any access would be filtered by `__getattribute__` (yes, the scary one) along the lines @richardjgowers suggested (`__getattr__` only gets called if there's no matching attribute). If updating is not set, proceed as normal. Otherwise check if a refresh is required before proceeding. This is the part that worries me the most, but I believe it can be made robust. (For normal, non-updating use it only adds one conditional checking layer);
- `AtomGroup` would need a sort of internal copy method, that changes it in-place to become another `AtomGroup`. If the state is only atom indices, then this is easy;
- The `AtomGroup` would also need to store a reference to the universe, if it doesn't already;
- Pickling/unpickling will be tricky with callbacks; I'd disable it for now, for those cases;
- Manual refresh could be an option (maybe by setting `AtomGroup.manual_update = True`).

A separate-class implementation would have pretty much the same elements, but in a different object, inheriting from `AtomGroup` as @richardjgowers suggested. The biggest differences are that users would need to keep track of which `AtomGroups` are updating and which are not, and non-updating `AtomGroups` couldn't be made updating.

**Comment by orbeckst** [comment]
Magic updating in one way or another would be great, preferably in a transparent and natural fashion.

It would really help if we managed to get 0.14.0 #646 out of the door ASAP and can merge #363 so that we don't start working on too many things in parallel. This discussion is a clear instance where we really should not be talking about old AtomGroup anymore. Instead, devs who are neither @richardjgowers or @dotsdl need to get used to the new hotness that is #363 in order to get a better understanding of what's going to work easily. 

tl;dr: **Let's ship 0.14.0!**

**Comment by mnmelo** [comment]
Not to worry, @orbeckst: these are just tentative API/implementation specs. When I finally write something I'll do it against the issue-363 branch (@richardgowers and @dotsdl, please let me know if you plan on doing any history rewriting there).
I also don't think there's any rush for this to be in 0.14.0, even if #363 is merged by then.

**Comment by richardjgowers** [comment]
@mnmelo yeah I always get getattr and getattribute mixed up.  But if we're filtering everything through getattribute, I'd rather we used a separate class for this so that we limit how much is exposed to this weird hack?

It also makes sense as they're kind of different objects, an AtomGroup is pointers to certain atoms, an AtomSelection is more like a description of the atoms it contains.. so the reprs would be like
```python
<AtomGroup with 100 atoms>

<AtomSelection of 'around 5.0 protein' currently 124 atoms>
```

WRT to be "future-proof", the new AtomGroup is just a np array of indices and a reference to Universe.  So modifying them in place should be fairly easy as long as the original selection string and original AtomGroup is stored.

**Comment by dotsdl** [comment]
@richardjgowers, @mnmelo because `AtomGroup` already has a good deal of metaclassing magic being done to construct it, I'd rather we have a separate object like `AtomSelection` that "has a" `AtomGroup` as a property that it applies the selection to. We can even duck-type it a bit by making this property `AtomSelection.atoms`, since you can always call any `*Group.atoms` and get back an `AtomGroup` representative of that `Group`. 

This makes the updating machinery explicit but also lets `AtomSelection` be used almost as normal. It also keeps `AtomGroup` thin and light for performance reasons, and limits how much state they carry around to only their identity. I think having `AtomGroup

**Comment by mnmelo** [comment]
@dotsdl, yup, I'll probably be going that way. As @jbarnoud pointed out (across the office) since we'll be hacking `__getattribute__` (or `__getattr__`, now that there's no attribute inheritance) we might as well have the `AtomGroup` as an attribute.

I asked for some advice on subclassing `AtomGroup` in #363. That may now be moot if we go the separate-class way, but I'm still curious to know how you envision that.
